### PR TITLE
CORE-627 COMM events metrics

### PR DESCRIPTION
**Pull request by ys-pyrofex**
## Overview
consume/produce/install COMM Events/seconds like we have done with produce and consume

### Does this PR relate to an RChain JIRA issue? 
https://rchain.atlassian.net/browse/CORE-627

### CODE REVIEW

File: rspace/src/main/scala/coop/rchain/rspace/package.scala [hunk 1]
```diff
@@ -141,17 +141,20 @@ package object rspace {
...
-              .foreach {
-                case DataCandidate(candidateChannel, Datum(_, persistData), dataIndex)
-                    if !persistData =>
-                  store.removeDatum(txn, Seq(candidateChannel), dataIndex)
-                case _ =>
-                  ()
-              }
-            logger.debug(s"consume: data found for <patterns: $patterns> at <channels: $channels>")
-            Some((continuation, dataCandidates.map(_.datum.a)))
+            store.eventsCounter.registerConsumeCommEvent {
```
Comment (henrytill) [inline rspace/src/main/scala/coop/rchain/rspace/package.scala]: It appears that you are measuring the duration of this block.  This isn't the right strategy - a COMM Event is not what happens inside of this block.  Rather it's the event of satisfying a match (or matches in the case of a join) above.  So the "event" from our perspective is when the call to `extractDataCandidates` has returned a `Some(...)`, so my comment before was to suggest that you increment the counter here.

In terms of a time-based metric, what we really need to be able to measure is how many times we are hitting this counter per second.  We also need to do this for produces and consumes.

**Comment by henrytill** [comment]
In addition to needing COMM Event per second measurements, I realize now that the we are still missing metrics for the number of produces per second and the number of consumes per second.  While I do think that the duration of produces and consumes is a useful metric, you'll notice that in the [ticket for CORE-529](https://rchain.atlassian.net/browse/CORE-529) it says:

> measure the number of consumes and produces that the tuplespace can handle in a second.

Because of the current locking semantics enforce a serial ordering, we could attempt to derive a {produce,consume}-per-second measurement from the current durations, but this seems problematic because it won't work as we begin to work towards a finer-grained concurrency model.  

Similarly, in the [ticket for CORE-629](https://rchain.atlassian.net/browse/CORE-529)

> We also need to calculate COMM Events/seconds

Apologies for not noticing and correcting this in #845.

File: rspace/src/main/scala/coop/rchain/rspace/RSpace.scala [hunk 2]
```diff
@@ -110,17 +110,20 @@ class RSpace[C, P, A, K](val store: IStore[C, P, A, K]) extends ISpace[C, P, A,
...
-              .foreach {
-                case DataCandidate(candidateChannel, Datum(_, persistData), dataIndex)
-                    if !persistData =>
-                  store.removeDatum(txn, Seq(candidateChannel), dataIndex)
-                case _ =>
-                  ()
-              }
-            logger.debug(s"consume: data found for <patterns: $patterns> at <channels: $channels>")
-            Some((continuation, dataCandidates.map(_.datum.a)))
+            store.eventsCounter.registerConsumeCommEvent {
```
Comment (henrytill) [inline rspace/src/main/scala/coop/rchain/rspace/RSpace.scala]: You are still measuring the duration of this block - **do not do this**.  This block _is not_ the COMM Event, so the duration of this block is meaningless - it is not the duration of a COMM Event.  **COMM Events do not have a duration**.  

As I said before, you should be only be incrementing the counter and the throughput statistics on this line.

File: rspace/src/main/scala/coop/rchain/rspace/RSpace.scala [hunk 3]
```diff
@@ -247,22 +252,24 @@ class RSpace[C, P, A, K](val store: IStore[C, P, A, K]) extends ISpace[C, P, A,
...
-              .sortBy(_.datumIndex)(Ordering[Int].reverse)
-              .foreach {
-                case DataCandidate(candidateChannel, Datum(_, persistData), dataIndex) =>
-                  if (!persistData && dataIndex >= 0) {
-                    store.removeDatum(txn, Seq(candidateChannel), dataIndex)
-                  }
-                  store.removeJoin(txn, candidateChannel, channels)
-                case _ =>
-                  ()
+            store.eventsCounter.registerProduceCommEvent {
```
Comment (henrytill) [inline rspace/src/main/scala/coop/rchain/rspace/RSpace.scala]: See above.

File: rspace/src/main/scala/coop/rchain/rspace/RSpace.scala [hunk 4]
```diff
@@ -163,15 +166,17 @@ class RSpace[C, P, A, K](val store: IStore[C, P, A, K]) extends ISpace[C, P, A,
                            |at <channels: $channels>""".stripMargin.replace('\n', ' '))
           None
         case Some(dataCandidates) =>
-          dataCandidates.foreach {
-            case DataCandidate(candidateChannel, Datum(_, persistData), dataIndex)
-                if !persistData =>
-              store.removeDatum(txn, Seq(candidateChannel), dataIndex)
-            case _ =>
-              ()
+          store.eventsCounter.registerInstallCommEvent {
```
Comment (henrytill) [inline rspace/src/main/scala/coop/rchain/rspace/RSpace.scala]: See above.

File: rspace/src/main/scala/coop/rchain/rspace/StoreEventsCounter.scala [hunk 5]
```diff
@@ -20,73 +27,163 @@ case class StoreCounters(sizeOnDisk: Long,
...
+
+  def registerConsume[T](f: => T): T =
+    register(consumesCounter, f)
+
+  private[rspace] def getConsumesCount: Long = consumesCounter.count
+
+  //consumes Comm
+  private[this] val consumesCommCounter = new Counter()
+
+  def registerConsumeCommEvent: Unit =
```
Comment (henrytill) [inline rspace/src/main/scala/coop/rchain/rspace/StoreEventsCounter.scala]: Let's add parens:
```scala
def registerConsumeCommEvent(): Unit
```

File: rspace/src/main/scala/coop/rchain/rspace/StoreEventsCounter.scala [hunk 6]
```diff
@@ -20,73 +27,163 @@ case class StoreCounters(sizeOnDisk: Long,
...
+    registerComm(consumesCommCounter)
+
+  private[rspace] def getConsumesCommCount: Long = consumesCommCounter.count
+
+  //produces Comm
+  private[this] val producesCommCounter = new Counter()
+
+  def registerProduceCommEvent: Unit =
```
Comment (henrytill) [inline rspace/src/main/scala/coop/rchain/rspace/StoreEventsCounter.scala]: Similarly:
```scala
def registerProduceCommEvent(): Unit
```

File: rspace/src/main/scala/coop/rchain/rspace/StoreEventsCounter.scala [hunk 7]
```diff
@@ -20,73 +27,163 @@ case class StoreCounters(sizeOnDisk: Long,
...
+    registerComm(producesCommCounter)
+
+  private[rspace] def getProducesCommCount: Long = producesCommCounter.count
+
+  //install Comm
+
+  private[this] val installCommCounter = new Counter()
+
+  def registerInstallCommEvent: Unit =
```
Comment (henrytill) [inline rspace/src/main/scala/coop/rchain/rspace/StoreEventsCounter.scala]: Similarly:
```scala
def registerInstallCommEvent(): Unit
```

File: node/src/main/scala/coop/rchain/node/diagnostics/StoreMetrics.scala [hunk 8]
```diff
@@ -28,15 +28,29 @@ object StoreMetrics extends StoreMetricsInstances {
...
       m.setGauge(name, value)
 
+    def cm(name: String, value: Option[StoreUsageCount]): F[Unit] =
+      for {
+        _ <- m.setGauge(name + "-count", value.map(_.count).getOrElse(0))
+        _ <- m.setGauge(name + "-peak-rate", value.map(_.peakRate.toLong).getOrElse(0))
+        _ <- m.setGauge(name + "-current-rate", value.map(_.currentRate.toLong).getOrElse(0))
+      } yield ()
+
+    def pc(name: String, value: Option[StoreUsageCount]): F[Unit] =
```
Comment (goral09) [inline node/src/main/scala/coop/rchain/node/diagnostics/StoreMetrics.scala]: Is this a shorthand for _produce_ ?

File: rspace/src/test/scala/coop/rchain/rspace/StorageActionsTests.scala [hunk 9]
```diff
@@ -268,6 +289,9 @@ trait StorageActionsTests extends StorageTestsBase[String, Pattern, String, Stri
 
     store.eventsCounter.getProducesCount shouldBe 2
     store.eventsCounter.getConsumesCount shouldBe 1
+    store.eventsCounter.getProducesCommCount shouldBe 1
+    store.eventsCounter.getConsumesCommCount shouldBe 0
```
Comment (goral09) [inline rspace/src/test/scala/coop/rchain/rspace/StorageActionsTests.scala]: I'm trying to understand how does _produce COMM event_ differs from _consume COMM event_. In the previous test you have: _producing and then consuming on the same channel_ which results in _getConsumesCommCount shouldBe 1_; here we have _producing on channel, consuming on that channel ... _ but _getConsumesCommCount == shouldBe 0 _. I don't understand that, the scenario is _almost_ the same as in the previous test case. Why is it that now there is only _produce COMM event_ ?

File: rspace/src/test/scala/coop/rchain/rspace/StorageActionsTests.scala [hunk 10]
```diff
@@ -377,6 +404,9 @@ trait StorageActionsTests extends StorageTestsBase[String, Pattern, String, Stri
 
     store.eventsCounter.getProducesCount shouldBe 3
     store.eventsCounter.getConsumesCount shouldBe 3
+    store.eventsCounter.getProducesCommCount shouldBe 0
+    store.eventsCounter.getConsumesCommCount shouldBe 3
+    store.eventsCounter.getInstallCommCount shouldBe 0
```
Comment (goral09) [inline rspace/src/test/scala/coop/rchain/rspace/StorageActionsTests.scala]: Isn't this the same as [this test](https://github.com/rchain/rchain/pull/914/files#diff-ef153c9df8e794c30aa19a89654b8185R189) ?

File: node/src/main/scala/coop/rchain/node/diagnostics/StoreMetrics.scala (hunk 8 above)
Comment (ys-pyrofex) [reply node/src/main/scala/coop/rchain/node/diagnostics/StoreMetrics.scala]: "pc" is abbreviation of produce/consume, and "cm" is for "comm".

File: rspace/src/test/scala/coop/rchain/rspace/StorageActionsTests.scala (hunk 9 above)
Comment (ys-pyrofex) [reply rspace/src/test/scala/coop/rchain/rspace/StorageActionsTests.scala]: Difference is in order of produce and consume calls. If we call consume before produce, than produce comm even occur. If we call produce and then consume - this is a consume comm event. If no data transferred (i.e. only produce, or only consume called) - there is no comm event.

File: rspace/src/test/scala/coop/rchain/rspace/StorageActionsTests.scala (hunk 10 above)
Comment (ys-pyrofex) [reply rspace/src/test/scala/coop/rchain/rspace/StorageActionsTests.scala]: See the explanation above.

**Pull request merged**